File: src/python/xerxes/cortex/agents/memory_integration.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Cortex memory integration for agent context and recall."""

from dataclasses import dataclass
from typing import Any

from ...memory import (
    ContextualMemory,
    EntityMemory,
    LongTermMemory,
    MemoryItem,
    ShortTermMemory,
    SQLiteStorage,
    UserMemory,
)


@dataclass
class CortexMemory:
    """Unified memory layer for Cortex agents.

    Combines short-term, long-term, entity, user, and contextual memory
    subsystems to provide rich context for agent tasks.

    Attributes:
        storage (SQLiteStorage | None): Persistent storage backend.
        short_term (ShortTermMemory | None): Recent interaction buffer.
        long_term (LongTermMemory | None): Durable memory store.
        entity_memory (EntityMemory | None): Entity tracking memory.
        user_memory (UserMemory | None): User-specific memory.
        contextual (ContextualMemory): Cross-memory search interface.
    """

    def __init__(
        self,
        enable_short_term: bool = True,
        enable_long_term: bool = True,
        enable_entity: bool = True,
        enable_user: bool = False,
        persistence_path: str | None = None,
        short_term_capacity: int = 50,
        long_term_capacity: int = 5000,
    ):
        """Build the four-tier memory stack for a Cortex run.

        Persistence is gated by the ``WRITE_MEMORY=1`` environment variable
        so tests and ephemeral runs never accidentally hit disk. When
        persistence is disabled the long-term, entity, and user memories
        still work — they just live in-memory.

        Args:
            enable_short_term: Allocate the recent-context buffer.
            enable_long_term: Allocate the durable knowledge store.
            enable_entity: Allocate the entity tracker.
            enable_user: Allocate per-user memory.
            persistence_path: SQLite database path used when persistence
                is enabled via ``WRITE_MEMORY``.
            short_term_capacity: Items kept in the short-term buffer.
            long_term_capacity: Hard cap on long-term store size.
        """

        import os

        write_memory = os.environ.get("WRITE_MEMORY", "0") == "1"

        self.storage = SQLiteStorage(persistence_path) if (persistence_path and write_memory) else None

        self.short_term = ShortTermMemory(capacity=short_term_capacity) if enable_short_term else None
        self.long_term = LongTermMemory(storage=self.storage, max_items=long_term_capacity) if enable_long_term else None
        self.entity_memory = EntityMemory(storage=self.storage) if enable_entity else None
        self.user_memory = UserMemory(storage=self.storage) if enable_user else None

        self.contextual = ContextualMemory(short_term_capacity=short_term_capacity, long_term_storage=self.storage)

    def build_context_for_task(
        self,
        task_description: str,
        agent_role: str | None = None,
        additional_context: str | None = None,
        max_items: int = 10,
    ) -> str:
        """Assemble a context string from every active memory layer.

        Pulls in (in order): supplied background, recent short-term items,
        agent-filtered long-term hits, extracted entities with their
        frequencies, and the top results from the contextual search.

        Args:
            task_description: Used as the search query and entity source.
            agent_role: Used to filter long-term retrieval by ``agent_id``.
            additional_context: Optional preamble shown verbatim.
            max_items: Cap on the contextual memory search.
        """

        context_parts = []

        if additional_context:
            context_parts.append(f"Background:\n{additional_context}")

        if self.short_term:
            recent = self.short_term.get_recent(n=5)
            if recent:
                context_parts.append("Recent context:")
                for item in recent:
                    context_parts.append(f"  • {item.content[:200]}")

        if self.long_term:
            relevant = self.long_term.search(
                query=task_description, limit=5, filters={"agent_id": agent_role} if agent_role else None
            )
            if relevant:
                context_parts.append("\nRelevant knowledge:")
                for item in relevant:
                    context_parts.append(f"  • {item.content[:200]}")

        if self.entity_memory:
            entities = self.entity_memory._extract_entities(task_description)
            if entities:
                context_parts.append("\nKnown entities:")
                for entity in entities[:5]:
                    info = self.entity_memory.get_entity_info(entity)
                    if info.get("frequency", 0) > 0:
                        context_parts.append(f"  • {entity}: {info.get('frequency')} mentions")

        comprehensive = self.contextual.search(query=task_description, limit=max_items, search_long_term=True)

        if comprehensive:
            context_parts.append("\nRelated memories:")
            for item in comprehensive[:3]:
                context_parts.append(f"  • {item.content[:150]}")

        return "\n".join(context_parts) if context_parts else ""

    def save_task_result(
        self,
        task_description: str,
        result: str,
        agent_role: str,
        importance: float = 0.5,
        task_metadata: dict[str, Any] | None = None,
    ):
        """Persist a completed task result across the memory tiers.

        Always writes to short-term, entity and contextual memory; long-term
        storage is only used when ``importance >= 0.7``.

        Args:
            task_description: Original task statement (stored truncated).
            result: Agent output to persist.
            agent_role: Tagged onto every memory entry for later filtering.
            importance: Score in ``[0, 1]`` gating long-term promotion.
            task_metadata: Extra metadata merged into the entry.
        """

        metadata = task_metadata or {}
        metadata["task"] = task_description[:100]
        metadata["agent_role"] = agent_role

        if self.short_term:
            self.short_term.save(
                content=f"Task completed: {task_description[:100]} - Result: {result[:200]}",
                metadata=metadata,
                agent_id=agent_role,
            )

        if self.long_term and importance >= 0.7:
            self.long_term.save(content=result, metadata=metadata, agent_id=agent_role, importance=importance)

        if self.entity_memory:
            self.entity_memory.save(content=f"{task_description} {result}", metadata=metadata)

        self.contextual.save(content=result, metadata=metadata, importance=importance, agent_id=agent_role)

    def save_agent_interaction(self, agent_role: str, action: str, content: str, importance: float = 0.3):
        """Log a one-line agent interaction (action + content).

        Always writes to short-term memory; promotes to long-term when
        ``importance >= 0.6``.
        """

        interaction = f"[{agent_role}] {action}: {content}"

        if self.short_term:
            self.short_term.save(content=interaction, metadata={"action": action}, agent_id=agent_role)

        if importance >= 0.6 and self.long_term:
            self.long_term.save(content=interaction, agent_id=agent_role, importance=importance)

    def save_cortex_decision(self, decision: str, context: str, outcome: str | None = None, importance: float = 0.7):
        """Record an orchestration-level decision under ``cortex_manager``.

        Writes to both long-term and contextual memory so the decision is
        searchable across future runs.
        """

        content = f"Decision: {decision}\nContext: {context}"
        if outcome:
            content += f"\nOutcome: {outcome}"

        metadata = {"type": "cortex_decision", "has_outcome": outcome is not None}

        if self.long_term:
            self.long_term.save(content=content, metadata=metadata, importance=importance, agent_id="cortex_manager")

        self.contextual.save(content=content, metadata=metadata, importance=importance)

    def get_agent_history(self, agent_role: str, limit: int = 20) -> list[str]:
        """Return up to ``limit`` content strings filtered to ``agent_role``.

        Short-term entries are returned first; the remainder of the budget
        is filled from long-term memory.
        """

        history = []

        if self.short_term:
            st_items = self.short_term.search(query="", limit=limit, filters={"agent_id": agent_role})
            history.extend([item.content for item in st_items])

        if self.long_term:
            lt_items = self.long_term.retrieve(filters={"agent_id": agent_role}, limit=max(0, limit - len(history)))
            if isinstance(lt_items, list):
                history.extend([item.content for item in lt_items])
            elif isinstance(lt_items, MemoryItem):
                history.append(lt_items.content)

        return history[:limit]

    def get_user_context(self, user_id: str) -> str:
        """Return ``UserMemory`` context for ``user_id`` or ``""`` if absent."""

        if self.user_memory:
            return self.user_memory.get_user_context(user_id)
        return ""

    def reset_short_term(self):
        """Clear the short-term memory buffer."""

        if self.short_term:
            self.short_term.clear()

    def reset_all(self):
        """Clear all memory subsystems."""

        if self.short_term:
            self.short_term.clear()
        if self.long_term:
            self.long_term.clear()
        if self.entity_memory:
            self.entity_memory.clear()
        if self.contextual:
            self.contextual.clear()

    def get_summary(self) -> str:
        """Return a human-readable summary across every active memory tier."""

        parts = []

        if self.short_term:
            parts.append(self.short_term.summarize())

        if self.long_term:
            parts.append(self.long_term.consolidate())

        if self.entity_memory:
            stats = self.entity_memory.get_statistics()
            if stats["total_items"] > 0:
                parts.append(f"Tracking {len(self.entity_memory.entities)} entities")

        return "\n\n".join(parts)

File: src/typescript/src/cortex/agents/memoryIntegration.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  ContextualMemory,
  EntityMemory,
  LongTermMemory,
  ShortTermMemory,
  SimpleStorage,
  UserMemory,
  type MemoryItem,
  type MemoryMetadata,
  type MemorySaveOptions,
  type MemoryStorage,
} from '../../memory/index.js'

/** Configuration for the Cortex-specific composition of the native memory tiers. */
export interface CortexMemoryOptions {
  readonly enableEntity?: boolean
  readonly enableLongTerm?: boolean
  readonly enableShortTerm?: boolean
  readonly enableUser?: boolean
  readonly longTermCapacity?: number
  /**
   * Explicit storage boundary. Omitting it intentionally selects ephemeral
   * in-process storage; this facade never reads WRITE_MEMORY or opens a path.
   */
  readonly storage?: MemoryStorage
  readonly shortTermCapacity?: number
}

export interface CortexMemoryContextOptions {
  readonly additionalContext?: string
  readonly agentRole?: string
  readonly maxItems?: number
}

export interface CortexTaskResultMemoryInput {
  readonly agentRole: string
  readonly importance?: number
  readonly metadata?: MemoryMetadata
  readonly result: string
  readonly taskDescription: string
}

export interface CortexAgentInteractionInput {
  readonly action: string
  readonly agentRole: string
  readonly content: string
  readonly importance?: number
}

export interface CortexDecisionMemoryInput {
  readonly context: string
  readonly decision: string
  readonly importance?: number
  readonly outcome?: string
}

/**
 * Unified native memory facade for Cortex agents.
 *
 * The individual tier implementations remain reusable on their own. This
 * class only composes them for Cortex task context and result persistence; it
 * does not create a database or consult process environment state by itself.
 */
export class CortexMemory {
  readonly contextual: ContextualMemory
  readonly entityMemory: EntityMemory | undefined
  readonly longTerm: LongTermMemory | undefined
  readonly shortTerm: ShortTermMemory | undefined
  readonly storage: MemoryStorage
  readonly userMemory: UserMemory | undefined

  constructor(options: CortexMemoryOptions = {}) {
    const shortTermCapacity = capacity(options.shortTermCapacity ?? 50, 'shortTermCapacity')
    const longTermCapacity = capacity(options.longTermCapacity ?? 5_000, 'longTermCapacity')
    this.storage = options.storage ?? new SimpleStorage()
    this.shortTerm = options.enableShortTerm === false
      ? undefined
      : new ShortTermMemory({ capacity: shortTermCapacity, storage: this.storage })
    this.longTerm = options.enableLongTerm === false
      ? undefined
      : new LongTermMemory({ enableEmbeddings: false, maxItems: longTermCapacity, storage: this.storage })
    this.entityMemory = options.enableEntity === false
      ? undefined
      : new EntityMemory({ storage: this.storage })
    this.userMemory = options.enableUser === true ? new UserMemory(this.storage) : undefined

    // Reuse the public long-term tier when present so Cortex searches and
    // orchestration writes observe one durable knowledge set instead of two
    // independently hydrated stores with the same storage keys.
    this.contextual = new ContextualMemory({
      shortTermCapacity,
      longTerm: this.longTerm ?? new LongTermMemory({
        enableEmbeddings: false,
        maxItems: longTermCapacity,
        storage: this.storage,
      }),
    })
  }

  /** Assemble bounded task context from active Cortex memory tiers. */
  buildContextForTask(taskDescription: string, options: CortexMemoryContextOptions = {}): string {
    const task = requiredText(taskDescription, 'taskDescription')
    const maxItems = nonNegativeInteger(options.maxItems ?? 10, 'maxItems')
    const parts: string[] = []
    const additional = options.additionalContext?.trim()
    if (additional) parts.push(`Background:\n${additional}`)

    const recent = this.shortTerm?.getRecent(5) ?? []
    if (recent.length) {
      parts.push('Recent context:')
      for (const item of recent) parts.push(`  • ${item.content.slice(0, 200)}`)
    }

    const agentRole = options.agentRole?.trim()
    const relevant = this.longTerm?.search(task, 5, agentRole ? { agentId: agentRole } : undefined) ?? []
    if (relevant.length) {
      parts.push('\nRelevant knowledge:')
      for (const item of relevant) parts.push(`  • ${item.content.slice(0, 200)}`)
    }

    const entities = this.entityMemory?.extractEntities(task) ?? []
    const known = entities.flatMap(entity => {
      const frequency = this.entityMemory?.getEntityInfo(entity).frequency ?? 0
      return frequency > 0 ? [`  • ${entity}: ${frequency} mentions`] : []
    }).slice(0, 5)
    if (known.length) parts.push('\nKnown entities:', ...known)

    const comprehensive = this.contextual.search(task, maxItems)
    if (comprehensive.length) {
      parts.push('\nRelated memories:')
      for (const item of comprehensive.slice(0, 3)) parts.push(`  • ${item.content.slice(0, 150)}`)
    }
    return parts.join('\n')
  }

  /** CortexOrchestrator-compatible writer backed by contextual memory. */
  save(content: string, metadata: MemoryMetadata = {}, options: MemorySaveOptions = {}): MemoryItem {
    return this.contextual.save(requiredText(content, 'content'), { ...metadata }, { ...options })
  }

  /** Persist an actual completed task across the relevant Cortex memory tiers. */
  saveTaskResult(input: CortexTaskResultMemoryInput): void {
    const taskDescription = requiredText(input.taskDescription, 'taskDescription')
    const result = requiredText(input.result, 'result')
    const agentRole = requiredText(input.agentRole, 'agentRole')
    const importance = importanceValue(input.importance ?? 0.5)
    const metadata: MemoryMetadata = {
      ...(input.metadata ?? {}),
      task: taskDescription.slice(0, 100),
      agent_role: agentRole,
    }

    this.shortTerm?.save(
      `Task completed: ${taskDescription.slice(0, 100)} - Result: ${result.slice(0, 200)}`,
      metadata,
      { agentId: agentRole },
    )
    this.entityMemory?.save(`${taskDescription} ${result}`, metadata, { agentId: agentRole })
    if (importance >= 0.7 && this.longTerm) {
      this.longTerm.save(result, metadata, { agentId: agentRole, importance })
      return
    }
    this.contextual.save(result, metadata, { agentId: agentRole, importance })
  }

  /** Record a concrete action and promote it only when its importance warrants it. */
  saveAgentInteraction(input: CortexAgentInteractionInput): void {
    const agentRole = requiredText(input.agentRole, 'agentRole')
    const action = requiredText(input.action, 'action')
    const content = requiredText(input.content, 'content')
    const importance = importanceValue(input.importance ?? 0.3)
    const interaction = `[${agentRole}] ${action}: ${content}`
    this.shortTerm?.save(interaction, { action }, { agentId: agentRole })
    if (importance >= 0.6) this.longTerm?.save(interaction, { action }, { agentId: agentRole, importance })
  }

  /** Record an orchestration decision in the shared contextual/long-term store. */
  saveCortexDecision(input: CortexDecisionMemoryInput): void {
    const decision = requiredText(input.decision, 'decision')
    const context = requiredText(input.context, 'context')
    const importance = importanceValue(input.importance ?? 0.7)
    const outcome = input.outcome?.trim()
    const content = [
      `Decision: ${decision}`,
      `Context: ${context}`,
      ...(outcome ? [`Outcome: ${outcome}`] : []),
    ].join('\n')
    const metadata = { type: 'cortex_decision', has_outcome: Boolean(outcome) }
    if (this.longTerm) {
      this.longTerm.save(content, metadata, { agentId: 'cortex_manager', importance })
      return
    }
    this.contextual.save(content, metadata, { agentId: 'cortex_manager', importance })
  }

  /** Return recent then durable entries belonging to one Cortex role. */
  getAgentHistory(agentRole: string, limit = 20): string[] {
    const role = requiredText(agentRole, 'agentRole')
    const count = nonNegativeInteger(limit, 'limit')
    const history = this.shortTerm?.search('', count, { agentId: role }).map(item => item.content) ?? []
    const remaining = Math.max(0, count - history.length)
    if (remaining) {
      const durable = this.longTerm?.retrieve(undefined, { agentId: role }, remaining)
      if (Array.isArray(durable)) history.push(...durable.map(item => item.content))
      else if (durable) history.push(durable.content)
    }
    return history.slice(0, count)
  }

  getUserContext(userId: string): string {
    return this.userMemory?.getUserContext(requiredText(userId, 'userId')) ?? ''
  }

  resetShortTerm(): void {
    this.shortTerm?.clear()
  }

  resetAll(): void {
    this.shortTerm?.clear()
    this.longTerm?.clear()
    this.entityMemory?.clear()
    this.contextual.clear()
  }

  /** Provide a compact cross-tier status summary, including durable-memory consolidation. */
  getSummary(): string {
    const parts: string[] = []
    if (this.shortTerm) parts.push(this.shortTerm.summarize())
    if (this.longTerm) parts.push(this.longTerm.consolidate())
    if (this.entityMemory && this.entityMemory.size > 0) {
      parts.push(`Tracking ${Object.keys(this.entityMemory.entities).length} entities`)
    }
    return parts.join('\n\n')
  }
}

function capacity(value: number, name: string): number {
  if (!Number.isInteger(value) || value < 1) throw new Error(`${name} must be a positive integer`)
  return value
}

function importanceValue(value: number): number {
  if (!Number.isFinite(value) || value < 0 || value > 1) throw new Error('importance must be between zero and one')
  return value
}

function nonNegativeInteger(value: number, name: string): number {
  if (!Number.isInteger(value) || value < 0) throw new Error(`${name} must be a non-negative integer`)
  return value
}

function requiredText(value: string, name: string): string {
  const normalized = value.trim()
  if (!normalized) throw new Error(`${name} cannot be empty`)
  return normalized
}
